library/Pdfexport/HeadlessChrome.php

```diff
@@ -247,106 +264,169 @@ public function fromHtml($html, $asFile = false)
+                            });
+
+                            return;
+                        }
 
-                $pdf = null;
-                $chrome->stderr->on('data', function ($chunk) use (&$pdf, $chrome, $loop, $killer) {
-                    Logger::debug('Caught browser output: %s', $chunk);
+                        Logger::warning('Failed to connect to remote chrome: %s (%s)', implode(':', $this->remote), $e);
+                        Logger::debug($e->getTraceAsString());
+                    }
 
-                    if (preg_match(self::DEBUG_ADDR_PATTERN, trim($chunk), $matches)) {
-                        $loop->cancelTimer($killer);
+                // Fallback to the local binary if a remote chrome is unavailable
+                case $this->binary !== null:
+                    $browserHome = $this->getFileStorage()->resolvePath('HOME');
+                    $commandLine = join(' ', [
+                        escapeshellarg($this->getBinary()),
+                        static::renderArgumentList([
+                            '--bwsi',
+                            '--headless',
+                            '--disable-gpu',
+                            '--no-sandbox',
+                            '--no-first-run',
+                            '--disable-dev-shm-usage',
+                            '--remote-debugging-port=0',
+                            '--homedir=' => $browserHome,
+                            '--user-data-dir=' => $browserHome
+                        ])
+                    ]);
+
+                    if (Platform::isLinux()) {
+                        Logger::debug('Starting browser process: HOME=%s exec %s', $browserHome, $commandLine);
+                        $chrome = new Process('exec ' . $commandLine, null, ['HOME' => $browserHome]);
+                    } else {
+                        Logger::debug('Starting browser process: %s', $commandLine);
+                        $chrome = new Process($commandLine);
+                    }
 
-                        try {
-                            $pdf = $this->printToPDF(
-                                $matches[1],
-                                $matches[2],
-                                ! $this->document->isEmpty() ? $this->document->getPrintParameters() : []
+                    $killer = Loop::addTimer(10, function (TimerInterface $timer) use ($chrome, $deferred) {
+                        $chrome->terminate(6); // SIGABRT
+
+                        $interval = $timer->getInterval();
+                        // Reject/resolve the promise on its own event tick not to block the current one.
+                        Loop::futureTick(function () use ($deferred, $interval) {
+                            $deferred->reject(
+                                new HeadlessChromeException(
+                                    sprintf(
+                                        'Browser timed out after %d seconds without the expected output',
```

Comment:
Log this, and reject the promise with `Received empty response or none at all from browser. Please check the logs for further details.`